### PR TITLE
client: fix handling of internal redirects

### PR DESCRIPTION
**Pull request by fscherf**
@mwx23 It seems that when i fixed the link bug you reported, i broke internal redirects. Could you test if this patch breaks your use case?

### CODE REVIEW

**Comment by maratori** [comment]
@fscherf Now we have ability to do e2e testing. Let's add a couple of tests for redirects and for links issue.

**Review by maratori — APPROVED** [review]

File: tests/test_redirects.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    context = await lona_app_context(setup_app)
+
+    async with async_playwright() as p:
+        browser = await p.chromium.launch()
+        browser_context = await browser.new_context()
+        page = await browser_context.new_page()
+
+        # test redirect to root
+        await page.goto(context.make_url('/redirect-to-root/'))
```
Comment (maratori) [inline tests/test_redirects.py:65]: Let's assert url as well
```suggestion
        await page.goto(context.make_url('/redirect-to-root/'))
        await page.wait_for_url('/')
```

File: tests/test_redirects.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    @app.route('/bar/')
+    class BarFooView(LonaView):
+        def handle_request(self, request):
+            nonlocal relative_redirected
+
+            if not relative_redirected:
+                relative_redirected = True
+
+                return {
+                    'redirect': '.',
```
Comment (maratori) [inline tests/test_redirects.py:42]: Let's add one more test with relative redirection `'redirect': './baz'`.

Also I have a question should `'redirect': './baz'` behave the same as  `'redirect': 'baz'`?

Comment (fscherf) [reply tests/test_redirects.py:42]: Ack!
Yes i think so.

Comment (maratori) [reply tests/test_redirects.py:42]: Asking because now it's not.

Comment (fscherf) [reply tests/test_redirects.py:42]: I saw it coming :D thanks for reporting, i'll fix it

**Comment by mwx23** [comment]
@fscherf I'm not sure it is working. The external link gets stripped to just the path when it gets sent back to the client.

Also on a side note, `target` seems to be ignored, but that's no big deal.

```python
from lona import LonaApp, LonaView
from lona.html import HTML, A, Br

app = LonaApp(__file__)


@app.route('/')
class MyView(LonaView):

    def handle_request(self, request):
        html = HTML(
            A(
                'internal link to /foo',
                href='/foo',
                # target='_blank',
            ),
            Br(),
            A(
                'external link to lona',
                href='https://github.com/lona-web-org/lona',
                # target='_blank',
            )
        )

        return html


@app.route('/foo')
class FooView(LonaView):

    def handle_request(self, request):
        html = HTML(
            '/foo'
        )

        return html


app.run(port=8080)
```

Here's browser console.log
```
lona tx >> lona:[1,null,101,["/",null]] lona.js:1858:17
lona rx << lona:[1,"10",204,null] lona.js:1867:17
lona rx << lona:[1,"10",203,[null,[502,[403,"23","",[[401,"19","a",[],[],{},{"href":"/foo"},[[402,"20","internal link to /foo"]]],[401,"24","br",[],[],{},{},[]],[401,"21","a",[],[],{},{"href":"https://github.com/lona-web-org/lona"},[[402,"22","external link to lona"]]]],{}]]]] lona.js:1867:17
lona rx << lona:[1,"10",205,null] lona.js:1867:17
lona tx >> lona:[1,null,101,["/lona-web-org/lona",null]] lona.js:1858:17
lona rx << lona:[1,"11",204,null] lona.js:1867:17
lona rx << lona:[1,"11",203,[null,[501,"<h1>404</h1>\n<p>Not found</p>"]]] lona.js:1867:17
lona rx << lona:[1,"11",205,null] lona.js:1867:17
```

Here's the `lona` debugging log 
```
/Users/mwx23/lona/venv/bin/python /Users/mwx23/lona/link_target_blank.py
MainThread                     DEBUG    11:26:11.557310 lona.server starting server in '/Users/mwx23/lona'
MainThread                     DEBUG    11:26:11.557577 lona.server setup settings
MainThread                     DEBUG    11:26:11.557613 lona.server loading settings from '/Users/mwx23/lona/lona/default_settings.py'
MainThread                     DEBUG    11:26:11.558113 lona.server applying settings post overrides
MainThread                     DEBUG    11:26:11.558155 lona.server setup templating
MainThread                     DEBUG    11:26:11.558195 lona.templating loading template_dirs '/var/folders/nz/gpbsm2b90ds20957t2ly6s5c0000gn/T/tmpxknmpntn/templates', '/Users/mwx23/lona/lona/templates'
MainThread                     DEBUG    11:26:11.558253 lona.server setup server state
MainThread                     DEBUG    11:26:11.558283 lona.server setup routing
MainThread                     DEBUG    11:26:11.558322 lona.server setup middleware controller
MainThread                     DEBUG    11:26:11.558349 lona.middlewares setup middlewares
MainThread                     DEBUG    11:26:11.558376 lona.middlewares loading lona.middlewares.sessions.LonaSessionMiddleware
MainThread                     DEBUG    11:26:11.558853 lona.middlewares loading lona.middlewares.lona_messages.LonaMessageMiddleware
MainThread                     DEBUG    11:26:11.559057 lona.middlewares discover middleware hooks
MainThread                     DEBUG    11:26:11.559094 lona.middlewares discovering <lona.middlewares.sessions.LonaSessionMiddleware object at 0x1047cbfa0>
MainThread                     DEBUG    11:26:11.559205 lona.middlewares <lona.middlewares.sessions.LonaSessionMiddleware object at 0x1047cbfa0>.on_startup discovered
MainThread                     DEBUG    11:26:11.559240 lona.middlewares <lona.middlewares.sessions.LonaSessionMiddleware object at 0x1047cbfa0>.handle_connection discovered
MainThread                     DEBUG    11:26:11.559267 lona.middlewares discovering <lona.middlewares.lona_messages.LonaMessageMiddleware object at 0x1047cbfd0>
MainThread                     DEBUG    11:26:11.559366 lona.middlewares <lona.middlewares.lona_messages.LonaMessageMiddleware object at 0x1047cbfd0>.handle_websocket_message discovered
MainThread                     DEBUG    11:26:11.559398 lona.server setup aiohttp routing
MainThread                     DEBUG    11:26:11.559425 lona.server static url set to '/static/{path:.*}'
MainThread                     DEBUG    11:26:11.559726 lona.server setup view loader
MainThread                     DEBUG    11:26:11.559764 lona.view_loader loading views from routes
MainThread                     DEBUG    11:26:11.559791 lona.view_loader loading <class '__main__.MyView'>
MainThread                     DEBUG    11:26:11.559939 lona.view_loader loading <class '__main__.FooView'>
MainThread                     DEBUG    11:26:11.559982 lona.view_loader loading lona.default_views.FrontendView
MainThread                     DEBUG    11:26:11.560223 lona.view_loader loading lona.default_views.Error403View
MainThread                     DEBUG    11:26:11.560281 lona.view_loader loading lona.default_views.Error404View
MainThread                     DEBUG    11:26:11.560327 lona.view_loader loading lona.default_views.Error500View
MainThread                     DEBUG    11:26:11.560370 lona.server setup response parser
MainThread                     DEBUG    11:26:11.560398 lona.server setup view runtime controller
MainThread                     DEBUG    11:26:11.560552 lona.client_pre_compiler pre compiling client
MainThread                     DEBUG    11:26:11.572360 lona.server setup static file
MainThread                     DEBUG    11:26:11.572415 lona.static_file_loader static dirs '/var/folders/nz/gpbsm2b90ds20957t2ly6s5c0000gn/T/tmpxknmpntn/static', '/Users/mwx23/lona/lona/static' loaded
MainThread                     DEBUG    11:26:11.572449 lona.static_file_loader discover node classes
MainThread                     DEBUG    11:26:11.572574 lona.static_file_loader 50 node classes discovered
MainThread                     DEBUG    11:26:11.572674 lona.static_file_loader 0 node stylesheets discovered
MainThread                     DEBUG    11:26:11.572701 lona.static_file_loader 0 node scripts discovered
MainThread                     DEBUG    11:26:11.572724 lona.static_file_loader rendering html files
MainThread                     DEBUG    11:26:11.572753 lona.templating searching for 'lona/style_tags.html'
MainThread                     DEBUG    11:26:11.573709 lona.templating 'lona/style_tags.html' is '/Users/mwx23/lona/lona/templates/lona/style_tags.html' 
MainThread                     DEBUG    11:26:11.573801 lona.templating searching for 'lona/script_tags.html'
MainThread                     DEBUG    11:26:11.574949 lona.templating 'lona/script_tags.html' is '/Users/mwx23/lona/lona/templates/lona/script_tags.html' 
MainThread                     DEBUG    11:26:11.575038 lona.server setup finish
MainThread                     DEBUG    11:26:11.575159 asyncio Using selector: KqueueSelector
MainThread                     DEBUG    11:26:11.575555 lona.server start
MainThread                     DEBUG    11:26:11.575603 lona.middlewares running on_startup with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>)>
MainThread                     DEBUG    11:26:11.575641 lona.middlewares running <lona.middlewares.sessions.LonaSessionMiddleware object at 0x1047cbfa0>.on_startup(<MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>)>)
======== Running on http://localhost:8080 ========
(Press CTRL+C to quit)
MainThread                     DEBUG    11:26:14.797021 lona.server.http http request incoming
MainThread                     DEBUG    11:26:14.797115 lona.server.http resolving path '/'
LonaWorker_0                   DEBUG    11:26:14.797256 lona.routing resolving '/'
LonaWorker_0                   DEBUG    11:26:14.797336 lona.routing <Route(/, <class '__main__.MyView'>)> matched
MainThread                     DEBUG    11:26:14.797678 lona.server.http route <Route(/, <class '__main__.MyView'>)> matched
LonaWorker_0                   DEBUG    11:26:14.797830 lona.middlewares running handle_connection with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, http_request=<Request GET / >, connection=<lona.connection.Connection object at 0x1048401c0>)>
LonaWorker_0                   DEBUG    11:26:14.797899 lona.middlewares running <lona.middlewares.sessions.LonaSessionMiddleware object at 0x1047cbfa0>.handle_connection(<MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, http_request=<Request GET / >, connection=<lona.connection.Connection object at 0x1048401c0>)>)
LonaWorker_0                   DEBUG    11:26:14.798049 lona.middlewares data was not handled
LonaRuntimeWorker_0            DEBUG    11:26:14.798497 lona.views trying to start a new view runtime
LonaRuntimeWorker_0            DEBUG    11:26:14.798561 lona.views running in frontend mode
LonaRuntimeWorker_0            DEBUG    11:26:14.798703 lona.middlewares running handle_request with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, connection=<lona.connection.Connection object at 0x1048401c0>, request=<lona.request.Request object at 0x10483c640>, view=<lona.default_views.FrontendView object at 0x10483c5b0>)>
LonaRuntimeWorker_0            DEBUG    11:26:14.798768 lona.middlewares data was not handled
LonaRuntimeWorker_0            DEBUG    11:26:14.798810 lona.views starting new view runtime
LonaRuntimeWorker_0            DEBUG    11:26:14.798892 lona.response_parser '<class 'lona.default_views.FrontendView'>' is a template view
LonaRuntimeWorker_0            DEBUG    11:26:14.798948 lona.templating searching for 'lona/frontend.html'
LonaRuntimeWorker_0            DEBUG    11:26:14.801327 lona.templating 'lona/frontend.html' is '/Users/mwx23/lona/lona/templates/lona/frontend.html' 
LonaRuntimeWorker_0            DEBUG    11:26:14.801496 lona.templating resolving static file path lona/style.css
LonaRuntimeWorker_0            DEBUG    11:26:14.801553 lona.static_file_loader resolving 'lona/style.css'
LonaRuntimeWorker_0            DEBUG    11:26:14.801602 lona.static_file_loader trying '/var/folders/nz/gpbsm2b90ds20957t2ly6s5c0000gn/T/tmpxknmpntn/static/lona/style.css'
LonaRuntimeWorker_0            DEBUG    11:26:14.801684 lona.static_file_loader trying '/Users/mwx23/lona/lona/static/lona/style.css'
LonaRuntimeWorker_0            DEBUG    11:26:14.801754 lona.static_file_loader returning '/Users/mwx23/lona/lona/static/lona/style.css'
MainThread                     INFO     11:26:14.803481 aiohttp.access 127.0.0.1 [27/Sep/2021:01:26:14 +0000] "GET / HTTP/1.1" 200 879 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:92.0) Gecko/20100101 Firefox/92.0"
LonaStaticWorker_0             DEBUG    11:26:14.830938 lona.static_file_loader resolving 'lona/lona.js'
LonaStaticWorker_0             DEBUG    11:26:14.831053 lona.static_file_loader returning javascript client
MainThread                     INFO     11:26:14.836188 aiohttp.access 127.0.0.1 [27/Sep/2021:01:26:14 +0000] "GET /static/lona/lona.js HTTP/1.1" 200 225 "http://localhost:8080/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:92.0) Gecko/20100101 Firefox/92.0"
MainThread                     DEBUG    11:26:14.847541 lona.server.http http request incoming
MainThread                     DEBUG    11:26:14.847729 lona.server.http resolving path '/'
MainThread                     DEBUG    11:26:14.848045 lona.server.http route <Route(/, <class '__main__.MyView'>)> matched
LonaWorker_0                   DEBUG    11:26:14.848412 lona.middlewares running handle_connection with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, http_request=<Request GET / >, connection=<lona.connection.Connection object at 0x1048d09d0>)>
LonaWorker_0                   DEBUG    11:26:14.848491 lona.middlewares running <lona.middlewares.sessions.LonaSessionMiddleware object at 0x1047cbfa0>.handle_connection(<MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, http_request=<Request GET / >, connection=<lona.connection.Connection object at 0x1048d09d0>)>)
LonaWorker_0                   DEBUG    11:26:14.848648 lona.middlewares data was not handled
MainThread                     DEBUG    11:26:14.849034 lona.server.websockets <lona.connection.Connection object at 0x1048d09d0> opened
MainThread                     DEBUG    11:26:14.868006 lona.server.websockets <lona.connection.Connection object at 0x1048d09d0> message received lona:[1,null,101,["/",null]]
LonaWorker_0                   DEBUG    11:26:14.868220 lona.middlewares running handle_websocket_message with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, connection=<lona.connection.Connection object at 0x1048d09d0>, message='lona:[1,null,101,["/",null]]')>
LonaWorker_0                   DEBUG    11:26:14.868300 lona.middlewares running <lona.middlewares.lona_messages.LonaMessageMiddleware object at 0x1047cbfd0>.handle_websocket_message(<MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, connection=<lona.connection.Connection object at 0x1048d09d0>, message='lona:[1,null,101,["/",null]]')>)
LonaWorker_0                   DEBUG    11:26:14.868447 lona.views view message decoded: ['/', None]
LonaWorker_0                   DEBUG    11:26:14.869786 lona.views removing old connections
LonaWorker_0                   DEBUG    11:26:14.869936 lona.views searching for running view runtime
LonaWorker_0                   DEBUG    11:26:14.869995 lona.views trying to start a new view runtime
LonaWorker_0                   DEBUG    11:26:14.870115 lona.middlewares running handle_request with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, connection=<lona.connection.Connection object at 0x1048d09d0>, request=<lona.request.Request object at 0x1048d06d0>, view=<__main__.MyView object at 0x1048d0730>)>
LonaWorker_0                   DEBUG    11:26:14.870177 lona.middlewares data was not handled
LonaWorker_0                   DEBUG    11:26:14.870217 lona.views starting new view runtime
LonaWorker_0                   DEBUG    11:26:14.870302 lona.views message handled
LonaWorker_0                   DEBUG    11:26:14.870345 lona.middlewares data got handled by <lona.middlewares.lona_messages.LonaMessageMiddleware object at 0x1047cbfd0>.handle_websocket_message
MainThread                     DEBUG    11:26:17.273396 lona.server.websockets <lona.connection.Connection object at 0x1048d09d0> message received lona:[1,null,101,["/lona-web-org/lona",null]]
LonaWorker_0                   DEBUG    11:26:17.273631 lona.middlewares running handle_websocket_message with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, connection=<lona.connection.Connection object at 0x1048d09d0>, message='lona:[1,null,101,["/lona-web-org/lona",null]]')>
LonaWorker_0                   DEBUG    11:26:17.273714 lona.middlewares running <lona.middlewares.lona_messages.LonaMessageMiddleware object at 0x1047cbfd0>.handle_websocket_message(<MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, connection=<lona.connection.Connection object at 0x1048d09d0>, message='lona:[1,null,101,["/lona-web-org/lona",null]]')>)
LonaWorker_0                   DEBUG    11:26:17.273823 lona.views view message decoded: ['/lona-web-org/lona', None]
LonaWorker_0                   DEBUG    11:26:17.273974 lona.routing resolving '/lona-web-org/lona'
LonaWorker_0                   DEBUG    11:26:17.274036 lona.routing no match for '/lona-web-org/lona'
LonaWorker_0                   DEBUG    11:26:17.274085 lona.views removing old connections
LonaWorker_0                   DEBUG    11:26:17.274194 lona.views searching for running view runtime
LonaWorker_0                   DEBUG    11:26:17.274279 lona.views trying to start a new view runtime
LonaWorker_0                   DEBUG    11:26:17.274481 lona.middlewares running handle_request with <MiddlewareData(server=<lona.server.LonaServer object at 0x1047d8460>, connection=<lona.connection.Connection object at 0x1048d09d0>, request=<lona.request.Request object at 0x10483c9a0>, view=<lona.default_views.Error404View object at 0x10483cb80>)>
LonaWorker_0                   DEBUG    11:26:17.274561 lona.middlewares data was not handled
LonaWorker_0                   DEBUG    11:26:17.274621 lona.views starting new view runtime
LonaWorker_0                   DEBUG    11:26:17.274764 lona.views message handled
LonaWorker_0                   DEBUG    11:26:17.274834 lona.middlewares data got handled by <lona.middlewares.lona_messages.LonaMessageMiddleware object at 0x1047cbfd0>.handle_websocket_message
LonaRuntimeWorker_0            DEBUG    11:26:17.275634 lona.response_parser '<class 'lona.default_views.Error404View'>' is a template view
LonaRuntimeWorker_0            DEBUG    11:26:17.275728 lona.templating searching for 'lona/404.html'
LonaRuntimeWorker_0            DEBUG    11:26:17.276541 lona.templating 'lona/404.html' is '/Users/mwx23/lona/lona/templates/lona/404.html' 

```

File: tests/test_redirects.py (hunk 2 above)
Comment (fscherf) [reply tests/test_redirects.py:62]: Fixed! And i added a test for this use case

**Comment by fscherf** [comment]
@mwx23 That's right! The client parses the given URL, removes the origin and only uses the pathname, hash and query string. To use external links, they have to be flagged as such explicitly. I am not sure if it should stay this way, but for now it's consistent with the behavior before i fixed it the first time.

**Comment by mwx23** [comment]
Sorry @fscherf I hadn't updated my test code to include the `interactive=False`. My bad! Yes, it does work fine 😁

**Comment by fscherf** [comment]
@mwx23 No worries! Thanks for testing :)

**Review by maratori — APPROVED** [review]
@fscherf great, thanks!

**Comment by fscherf** [comment]
@mwx23, @maratori Thanks for reviewing!

**Pull request merged**